src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceCollection.cs

```diff
@@ -81,13 +103,117 @@ public int IndexOf(ServiceDescriptor item)
         /// <inheritdoc />
         public void Insert(int index, ServiceDescriptor item)
         {
+            // In theory, we should preserve the relative order here but it's expensive to find that out
+            // and we handle operations that rely on the relative order already
+            AddToServiceTypeMap(item);
+
             _descriptors.Insert(index, item);
         }
 
         /// <inheritdoc />
         public void RemoveAt(int index)
         {
+            // This will throw if the index is out of range
+            ServiceDescriptor item = _descriptors[index];
+            if (_serviceTypes.TryGetValue(item.ServiceType, out List<ServiceDescriptor> items))
+            {
+                items.Remove(item);
+            }
+
             _descriptors.RemoveAt(index);
         }
+
+        internal bool TryAdd(ServiceDescriptor item)
+        {
+            if (!_serviceTypes.TryGetValue(item.ServiceType, out List<ServiceDescriptor> items))
+            {
+                _serviceTypes[item.ServiceType] = new List<ServiceDescriptor> { item };
+                _descriptors.Add(item);
+                return true;
+            }
+            return false;
```

Comment:
When learn these code, just found here is a bug maybe?

Based on current implementation of 'Remove' related [methods](https://github.com/dotnet/runtime/blob/79c0e0417ad98ac6021303a1a11180c1be8012ab/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceCollection.cs#L75), they are not removing empty collection for specific ServiceType from dictionary if there is only one item for that ServiceType before calling 'Remove()'. So when user calls TryAdd() here after calling Remove(), it will refuse to add requested item with previous ServiceType even if it is not existing actually.

Add serviceDescripter (TypeA) -> Remove (TypeA) -> TryAdd() (TypeA) (will fail)

Reply:
That's a bug in the current PR, that should be fixed! Good catch!